crates/data_components/src/arrow/write.rs

```diff
@@ -955,16 +953,38 @@ impl DataSink for MemSink {
 
         // Ensure new data has no primary key conflicts internally, and generate primary key ids for later comparison to existing partition data.
         // We must also check for null values in primary keys. With that we can safely assume [`self.batches`] has no null primary keys.
+        //
+        // For InsertOp::Replace, we allow duplicate primary keys in new data because the operation will:
```

Comment:
Wouldn't the correct fix just to not add primary keys to the caching table? It feels incorrect to relax this constraint check